File: src/main/java/com/now/nowbot/service/messageServiceImpl/MapPoolService.java
package com.now.nowbot.service.messageServiceImpl;

import com.fasterxml.jackson.databind.JsonNode;
import com.now.nowbot.config.NowbotConfig;
import com.now.nowbot.model.enums.OsuMode;
import com.now.nowbot.model.mappool.now.Pool;
import com.now.nowbot.model.mappool.old.MapPoolDto;
import com.now.nowbot.qq.event.MessageEvent;
import com.now.nowbot.service.ImageService;
import com.now.nowbot.service.MessageService;
import com.now.nowbot.service.osuApiService.OsuBeatmapApiService;
import com.now.nowbot.throwable.TipsException;
import com.now.nowbot.util.ASyncMessageUtil;
import com.now.nowbot.util.CmdUtil;
import com.now.nowbot.util.Instruction;
import com.now.nowbot.util.JacksonUtil;
import jakarta.annotation.Resource;
import org.springframework.stereotype.Service;
import org.springframework.util.StringUtils;
import org.springframework.web.client.HttpClientErrorException;
import org.springframework.web.reactive.function.client.WebClient;
import org.springframework.web.reactive.function.client.WebClientResponseException;
import org.springframework.web.util.UriComponentsBuilder;

import java.time.Duration;
import java.util.List;
import java.util.Optional;

import static com.now.nowbot.util.command.CommandPatternStaticKt.FLAG_NAME;

@Service("MAP_POOL")
public class MapPoolService implements MessageService<MapPoolService.PoolParam> {

    private static final String           api   = NowbotConfig.BS_API_URL;
    private static final Optional<String> token = NowbotConfig.BS_TOKEN;
    @Resource
    ImageService         imageService;
    @Resource
    OsuBeatmapApiService osuBeatmapApiService;
    @Resource
    WebClient            webClient;

    public record PoolParam(int id, String name, OsuMode mode) {}

    @Override
    public boolean isHandle(MessageEvent event, String messageText, DataValue<PoolParam> data) throws TipsException {
        var m = Instruction.MAP_POOL.matcher(messageText);
        if (!m.find()) return false;

        String name = m.group(FLAG_NAME);
        OsuMode mode = CmdUtil.getMode(m).getData();

        if (! StringUtils.hasText(name)) {
            throw new TipsException("id 解析错误, 请确保只有数字");
        }

        try {
            int id = Integer.parseInt(name);
            data.setValue(new PoolParam(id, null, mode));
        } catch (NumberFormatException e) {
            data.setValue(new PoolParam(0, name, mode));
        }

        return true;
    }

    @Override
    public void HandleMessage(MessageEvent event, PoolParam param) throws Throwable {
        byte[] image;
        if (StringUtils.hasText(param.name())) {
            var result = searchByName(param.name());
            if (result.isEmpty())
                throw new TipsException(STR."未找到名称包含 \{param.name()} 的图池");
            if (result.size() == 1) {
                image = imageService.getPanelH(new MapPoolDto(result.getFirst(), osuBeatmapApiService), param.mode());
            } else {
                StringBuilder sb = new StringBuilder("查到了多个图池, 请确认结果:\n");
                for (int i = 0; i < result.size(); i++) {
                    sb.append(i + 1).append(": ").append(result.get(i).getName()).append('\n');
                }
                sb.append("p.s. 请直接发送选项对应的数字");

                var image2 = imageService.getPanelAlpha(sb);
                event.reply(image2);

                var lock = ASyncMessageUtil.getLock(event);
                var newEvent = lock.get();
                int n = 0;
                try {
                    n = Integer.parseInt(newEvent.getRawMessage());
                } catch (NumberFormatException e) {
                    throw new TipsException("输入错误");
                }
                if (n < 1 || n > result.size()) throw new TipsException("输入错误");

                image = imageService.getPanelH(new MapPoolDto(result.get(n - 1), osuBeatmapApiService), param.mode());
            }
        } else {
            var p = searchById(param.id());
            image = imageService.getPanelH(p.map(pool -> new MapPoolDto(pool, osuBeatmapApiService)).orElseThrow(() -> new TipsException(STR."未找到id为 \{param.id()} 的图池")), param.mode());
        }

        event.reply(image);
    }

    public List<Pool> searchByName(String name) {
        var nodeOpt = webClient.get()
                .uri(u -> UriComponentsBuilder.fromHttpUrl(api).path("/api/public/searchPool").queryParam("poolName", name).build().toUri())
                .headers(h -> token.ifPresent(t -> h.addIfAbsent("AuthorizationX", t)))
                .retrieve()
                .bodyToMono(JsonNode.class)
                .blockOptional(Duration.ofSeconds(30));
        return nodeOpt.map(node -> JacksonUtil.parseObjectList(node.get("data"), Pool.class)).orElseThrow();
    }

    public Optional<Pool> searchById(int id) {
        try {
            return webClient.get()
                    .uri(u -> UriComponentsBuilder.fromHttpUrl(api).path("/api/public/searchPool").queryParam("poolId", id).build().toUri())
                    .headers(h -> token.ifPresent(t -> h.addIfAbsent("AuthorizationX", t)))
                    .retrieve()
                    .bodyToMono(JsonNode.class)
                    .<Optional<Pool>>map(json -> json.has("data") ? Optional.ofNullable(JacksonUtil.parseObject(json.get("data"), Pool.class)) : Optional.empty())
                    .block(Duration.ofSeconds(30));
        } catch (HttpClientErrorException.NotFound | WebClientResponseException.NotFound e) {
            return Optional.empty();
        }
    }
}

File: src/main/java/com/now/nowbot/service/messageServiceImpl/MapPoolService.kt
package com.now.nowbot.service.messageServiceImpl

import com.fasterxml.jackson.databind.JsonNode
import com.now.nowbot.config.BeatmapMirrorConfig
import com.now.nowbot.model.enums.OsuMode
import com.now.nowbot.model.mappool.now.Pool
import com.now.nowbot.model.mappool.old.MapPoolDto
import com.now.nowbot.qq.event.MessageEvent
import com.now.nowbot.service.ImageService
import com.now.nowbot.service.MessageService
import com.now.nowbot.service.MessageService.DataValue
import com.now.nowbot.service.messageServiceImpl.MapPoolService.PoolParam
import com.now.nowbot.service.osuApiService.OsuBeatmapApiService
import com.now.nowbot.service.osuApiService.OsuCalculateApiService
import com.now.nowbot.throwable.TipsException
import com.now.nowbot.util.ASyncMessageUtil
import com.now.nowbot.util.CmdUtil.getMode
import com.now.nowbot.util.Instruction
import com.now.nowbot.util.JacksonUtil
import com.now.nowbot.util.command.FLAG_NAME
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service
import org.springframework.web.client.HttpClientErrorException
import org.springframework.web.reactive.function.client.WebClient
import org.springframework.web.reactive.function.client.WebClientResponseException
import org.springframework.web.util.UriComponentsBuilder
import java.time.Duration
import java.util.*

@Service("MAP_POOL") class MapPoolService(
    private val imageService: ImageService,
    private val beatmapApiService: OsuBeatmapApiService,
    private val calculateApiService: OsuCalculateApiService,
    private val webClient: WebClient,
    beatmapMirrorConfig: BeatmapMirrorConfig
) : MessageService<PoolParam> {
    private val url = beatmapMirrorConfig.url
    private val token = beatmapMirrorConfig.token

    data class PoolParam(val id: Int, val name: String?, val mode: OsuMode)

    @Throws(TipsException::class) override fun isHandle(
        event: MessageEvent,
        messageText: String,
        data: DataValue<PoolParam>,
    ): Boolean {
        val m = Instruction.MAP_POOL.matcher(messageText)
        if (!m.find()) return false

        val name: String? = m.group(FLAG_NAME)
        val mode = getMode(m).data!!

        if (name.isNullOrBlank()) {
            throw TipsException("请输入正确的图池名称！")
        }

        val id = name.toIntOrNull()

        if (id != null) {
            data.value = PoolParam(id, null, mode)
        } else {
            data.value = PoolParam(0, name, mode)
        }

        return true
    }

    @Throws(Throwable::class) override fun HandleMessage(event: MessageEvent, param: PoolParam) {
        if (param.name.isNullOrBlank().not()) {
            val result = searchByName(param.name!!)
            if (result.isEmpty()) throw TipsException("未找到名称包含 ${param.name} 的图池")

            if (result.size == 1) {
                event.reply(MapPoolDto(result.first(), beatmapApiService, calculateApiService), param.mode)
            } else {
                val sb = StringBuilder("查到了多个图池, 请确认结果:\n")
                for (i in result.indices) {
                    sb.append(i + 1).append(": ").append(result[i].name).append('\n')
                }
                sb.append("p.s. 请直接发送选项对应的数字")

                val image2 = imageService.getPanelAlpha(sb)
                event.reply(image2)

                val lock = ASyncMessageUtil.getLock(event)
                val newEvent = lock.get()

                val n: Int = newEvent.rawMessage.toIntOrNull() ?: throw TipsException("输入错误")

                if (n !in 1..result.size) throw TipsException("输入错误")

                event.reply(MapPoolDto(result[n - 1], beatmapApiService, calculateApiService), param.mode)
            }
        } else {
            val pool = searchByID(param.id) ?: throw TipsException("未找到id为 ${param.id} 的图池")

            event.reply(MapPoolDto(pool, beatmapApiService, calculateApiService), param.mode)
        }
    }

    private fun MessageEvent.reply(data: MapPoolDto, mode: OsuMode) {
        val body = mapOf("pool" to data, "mode" to mode.shortName)

        val image = imageService.getPanel(body, "H")

        this.reply(image)
    }

    fun searchByName(name: String): List<Pool> {
        if (url == null) return emptyList()

        val nodeOpt = webClient.get().uri {
            UriComponentsBuilder.fromUriString(url).path("/api/public/searchPool").queryParam("poolName", name).build()
                .toUri()
        }.headers {
            it.addIfAbsent("AuthorizationX", token)
        }.retrieve().bodyToMono(JsonNode::class.java).block(Duration.ofSeconds(30)) ?: return emptyList()

        return nodeOpt.map { node: JsonNode ->
            JacksonUtil.parseObjectList(node["data"], Pool::class.java)
        }.flatten()
    }

    fun searchByID(id: Int): Pool? {
        if (url == null) return null

        return try {
            webClient.get().uri {
                UriComponentsBuilder.fromUriString(url).path("/api/public/searchPool").queryParam("poolId", id).build()
                    .toUri()
            }.headers {
                it.addIfAbsent("AuthorizationX", token)
            }.retrieve().bodyToMono(JsonNode::class.java).map { json: JsonNode ->
                if (json.has("data")) JacksonUtil.parseObject(json["data"], Pool::class.java)
                else null
            }.block(Duration.ofSeconds(30))
        } catch (e: HttpClientErrorException.NotFound) {
            null
        } catch (e: WebClientResponseException.NotFound) {
            null
        }
    }

    companion object {
        private val log: Logger = LoggerFactory.getLogger(MapPoolService::class.java)
    }
}
